### PR TITLE
[Profiler] Use Exec to avoid templating issues.

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?

* Prevents shell injection via the 'format' parameter.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/reporter/profile_manager.py [hunk 1]
```diff
@@ -49,33 +49,46 @@ def _format_failed_pyspy_command(cmd, stdout, stderr) -> str:
...
+async def _get_pyspy_path() -> str:
+    process = await asyncio.create_subprocess_exec(
+        "which", "py-spy",
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    stdout, _ = await process.communicate()
+    if process.returncode != 0:
+        return ""
+    return stdout.decode().strip()
```
Comment (thomasdesr) [inline dashboard/modules/reporter/profile_manager.py:70]: Can this be? https://docs.python.org/3/library/shutil.html#shutil.which

File: dashboard/modules/reporter/profile_manager.py [hunk 2]
```diff
@@ -49,33 +49,46 @@ def _format_failed_pyspy_command(cmd, stdout, stderr) -> str:
...
-            cmd += " --native"
+            cmd.append("--native")
         if await _can_passwordless_sudo():
-            cmd = "sudo -n " + cmd.replace("py-spy", "$(which py-spy)")
-        process = await asyncio.create_subprocess_shell(
-            cmd,
+            pyspy_path = await _get_pyspy_path()
+            if pyspy_path != "":
+                cmd[0] = pyspy_path
+                cmd = ["sudo", "-n"] + cmd
```
Comment (thomasdesr) [inline dashboard/modules/reporter/profile_manager.py:87]: Why set pyspy's path dynamically here vs in the original cmd construction?

I assume its trying to preserving existing behavior? Any reason simplifying isn't a good idea?

Comment (ijrsvt) [reply dashboard/modules/reporter/profile_manager.py:87]: Ooh, I'll switch to shutil.which and always use that.

File: dashboard/modules/reporter/profile_manager.py (hunk 1 above)
Comment (ijrsvt) [reply dashboard/modules/reporter/profile_manager.py]: Done! Great lib!

**Review by rkooo567 — APPROVED** [review]
lol good catch..

**Comment by ijrsvt** [comment]
Failing tests are not related to the PR. Tested that profiling still works locally.

**Pull request merged**